tests/integration/test_observability.py

```diff
@@ -0,0 +1,389 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Limited
+# See LICENSE file for licensing details.
+
+
+import asyncio
+import datetime
+import json
+import logging
+import subprocess
+import urllib.request
+from pathlib import Path
+
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+from tenacity import retry, stop_after_attempt, wait_fixed
+
+from .helpers import (
+    BUCKET_NAME,
+    add_juju_secret,
+    fetch_action_sync_s3_credentials,
+    get_secret_data,
+    juju_sleep,
+    setup_s3_bucket_for_sch_server,
+)
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+CONTAINER_NAME = "test-container"
+SECRET_NAME_PREFIX = "integrator-hub-conf-"
+
+
+@retry(
+    wait=wait_fixed(5),
+    stop=stop_after_attempt(120),
+    reraise=True,
+)
+def check_metrics(address: str) -> None:
+    metrics = json.loads(urllib.request.urlopen(f"http://{address}:9091/api/v1/metrics").read())
+
+    logger.info(f"Metrics: {metrics} at time: {datetime.datetime.now()}")
+
+    assert len(metrics["data"]) > 0
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(
+    ops_test: OpsTest, charm_versions, azure_credentials, service_account, namespace
+):
+    """Build the charm-under-test and deploy it together with related charms.
+
+    Assert on the unit status before any relations/configurations take place.
+    """
+    logger.info("Setting up minio.....")
+
+    setup_minio_output = (
+        subprocess.check_output(
+            "./tests/integration/setup/setup_minio.sh | tail -n 1", shell=True, stderr=None
+        )
+        .decode("utf-8")
+        .strip()
+    )
+
+    logger.info(f"Minio output:\n{setup_minio_output}")
+
+    s3_params = setup_minio_output.strip().split(",")
+    endpoint_url = s3_params[0]
+    access_key = s3_params[1]
+    secret_key = s3_params[2]
+
+    logger.info(
+        f"Setting up s3 bucket with endpoint_url={endpoint_url}, access_key={access_key}, secret_key={secret_key}"
+    )
+
+    setup_s3_bucket_for_sch_server(endpoint_url, access_key, secret_key)
+
+    logger.info("Bucket setup complete")
+
+    logger.info("Building charm")
+    # Build and deploy charm from local source folder
+
+    charm = await ops_test.build_charm(".")
+
+    image_version = METADATA["resources"]["integration-hub-image"]["upstream-source"]
+
+    logger.info(f"Image version: {image_version}")
+
+    resources = {"integration-hub-image": image_version}
+
+    logger.info(
+        "Deploying Spark Integration hub charm, s3-integrator charm and azure-storage-integrator charm"
+    )
+
+    # Deploy the charm and wait for waiting status
+    await asyncio.gather(
+        ops_test.model.deploy(**charm_versions.s3.deploy_dict()),
+        ops_test.model.deploy(**charm_versions.azure_storage.deploy_dict()),
```

Comment:
Here I don't understand the need for deploying both Azure and S3 object storage backends -- is this required for the test? Also, can't we use the `deploy_minimal_kyuubi_setup` helper that sets kyuubi + s3 + hub up until a running Kyuubi setup.

Reply:
I cleanup the test, thanks for spotting